### DISCUSSION TITLE
Adding a custom provider for MongoDB [v6]

### DISCUSSION THREAD

**Discussion by usersina** [adonisjs/core#4802] [Help]
No mongoose or anything, I just want to add a simple MongoDB provider and I actually managed to do it.

However, I get typescript errors and it's not type-safe.

```typescript
// adonisrc.ts

...
  providers: [
    ...
    () => import('#providers/mongo_provider'),
  ],
...
```

```typescript
// providers/mongo_provider.ts

import type { ApplicationService } from '@adonisjs/core/types'
import { MongoClient } from 'mongodb'

import env from '#start/env'

export default class MongoProvider {
  constructor(protected app: ApplicationService) {}

  /**
   * Register bindings to the container
   */
  register() {
    this.app.container.singleton('mongo', () => { // <-- typescript error "No overload matches this call."
      return new MongoClient(env.get('MONGO_URI'))
    })
  }

  /**
   * The container bindings have booted
   */
  async boot() {
    const mongoClient = await this.app.container.make('mongo') // <-- `mongoClient` is not typesafe
    await mongoClient.connect()
  }

  /**
   * The application has been booted
   */
  async start() {
    const mongoClient = await this.app.container.make('mongo') // <-- `mongoClient` is not typesafe
    await mongoClient.db('admin').command({ ping: 1 })
  }

  /**
   * The process has been started
   */
  async ready() {}

  /**
   * Preparing to shutdown the app
   */
  async shutdown() {
    const mongoClient = await this.app.container.make('mongo') // <-- `mongoClient` is not typesafe
    await mongoClient.close()
  }
}
```

Needless to say, `npm run typecheck` fails. Here's the typescript error:
> Argument of type `String` is not assignable to parameter of type `AbstractConstructor<any>`

> No overload matches this call.
  Overload 1 of 2, '(binding: keyof ContainerBindings, resolver: BindingResolver<ContainerBindings, ApplicationService | Kernel | Dumper | ... 8 more ... | Repl>): void', gave the following error.
    Argument of type '"mongo"' is not assignable to parameter of type 'keyof ContainerBindings'.
  Overload 2 of 2, '(binding: AbstractConstructor<any>, resolver: BindingResolver<ContainerBindings, any>): void', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'AbstractConstructor<any>'.

**Comment by usersina** [comment]
I figured it out reading through the [`Dependency injection (DI)` -> `Defining static types for bindings`](https://docs.adonisjs.com/guides/concepts/dependency-injection#defining-static-types-for-bindings) section in the documentation.

```typescript
// app/types/container.ts

import { MongoClient } from 'mongodb'

declare module '@adonisjs/core/types' {
  interface ContainerBindings {
    mongo: MongoClient
  }
}
```

**Reply by pk4all** [reply]
How to use Auth provider using mongodb?
